sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/queries/AsyncSelectQueryTypeTest.kt

```diff
@@ -0,0 +1,154 @@
+import app.cash.sqldelight.core.compiler.SelectQueryGenerator
+import app.cash.sqldelight.core.dialects.intType
+import app.cash.sqldelight.core.dialects.textType
+import app.cash.sqldelight.dialects.postgresql.PostgreSqlDialect
+import app.cash.sqldelight.test.util.FixtureCompiler
+import com.google.common.truth.Truth.assertThat
+import com.squareup.burst.BurstJUnit4
+import org.junit.Assume.assumeTrue
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TemporaryFolder
+import org.junit.runner.RunWith
+
+@RunWith(BurstJUnit4::class)
+class AsyncSelectQueryTypeTest {
+  @get:Rule val tempFolder = TemporaryFolder()
+
+  @Test fun `returning clause correctly generates an async query function`(dialect: TestDialect) {
+    assumeTrue(dialect in listOf(TestDialect.POSTGRESQL, TestDialect.SQLITE_3_35))
+    val file = FixtureCompiler.parseSql(
+      """
+      |CREATE TABLE data (
+      |  val1 TEXT,
+      |  val2 TEXT
+      |);
+      |
+      |insertReturning:
+      |INSERT INTO data
+      |VALUES ('sup', 'dude')
+      |RETURNING *;
+      |""".trimMargin(),
+      tempFolder,
+      dialect = PostgreSqlDialect(),
+      generateAsync = true
+    )
+
+    val query = file.namedQueries.first()
+    val generator = SelectQueryGenerator(query)
+
+    assertThat(generator.customResultTypeFunction().toString()).isEqualTo(
+      """
+      |public suspend fun <T : kotlin.Any> insertReturning(mapper: (val1: kotlin.String?, val2: kotlin.String?) -> T): app.cash.sqldelight.async.AsyncExecutableQuery<T> = app.cash.sqldelight.async.AsyncQuery(${query.id}, driver, "Test.sq", "insertReturning", ""${'"'}
+      ||INSERT INTO data
+      ||VALUES ('sup', 'dude')
+      ||RETURNING *
+      |""${'"'}.trimMargin()) { cursor ->
+      |  check(cursor is app.cash.sqldelight.driver.r2dbc.R2dbcCursor)
+      |  mapper(
+      |    cursor.getString(0),
+      |    cursor.getString(1)
+      |  )
+      |}
+      |""".trimMargin()
+    )
+  }
+
+  @Test fun `returning clause in an update correctly generates an async query function`(dialect: TestDialect) {
+    assumeTrue(dialect in listOf(TestDialect.POSTGRESQL, TestDialect.SQLITE_3_35))
+    val file = FixtureCompiler.parseSql(
+      """
+      |CREATE TABLE IF NOT EXISTS users(
+      |    id ${dialect.intType} PRIMARY KEY,
+      |    firstname ${dialect.textType} NOT NULL,
+      |    lastname ${dialect.textType} NOT NULL
+      |);
+      |
+      |update:
+      |UPDATE users SET
+      |    firstname = :firstname,
+      |    lastname = :lastname
+      |WHERE id = :id
+      |RETURNING id, firstname, lastname;
+      |""".trimMargin(),
+      tempFolder,
+      dialect = PostgreSqlDialect(),
+      generateAsync = true
+    )
+
+    val query = file.namedQueries.first()
+    val generator = SelectQueryGenerator(query)
+
+    assertThat(generator.customResultTypeFunction().toString()).isEqualTo(
+      """
+      |public suspend fun <T : kotlin.Any> update(
```

Comment:
do these functions need to be suspending? I guess it doesn't really matter since you almost always execute them in the same code block

Reply:
Oh actually you're right, these shouldn't be suspending. The suspending call would be made on the returned `UpdateQuery`. For correctness-sake I'll remove the modifier there.